Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccticketing/ui/certificateselection/DccTicketingCertificateSelectionFragment.kt

```diff
@@ -1,24 +1,120 @@
-import android.view.LayoutInflater
 import android.view.View
-import android.view.ViewGroup
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.FragmentManager
+import androidx.lifecycle.lifecycleScope
 import androidx.navigation.fragment.navArgs
 import de.rki.coronawarnapp.R
+import de.rki.coronawarnapp.databinding.FragmentDccTicketingCertificateSelectionBinding
+import de.rki.coronawarnapp.util.di.AutoInject
+import de.rki.coronawarnapp.util.lists.decorations.TopBottomPaddingDecorator
+import de.rki.coronawarnapp.util.lists.diffutil.update
+import de.rki.coronawarnapp.util.ui.doNavigate
+import de.rki.coronawarnapp.util.ui.popBackStack
+import de.rki.coronawarnapp.util.ui.viewBinding
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModelFactoryProvider
+import de.rki.coronawarnapp.util.viewmodel.cwaViewModelsAssisted
+import setTextWithUrl
+import java.util.Locale
+import javax.inject.Inject
 
-class DccTicketingCertificateSelectionFragment : Fragment() {
+class DccTicketingCertificateSelectionFragment :
+    Fragment(R.layout.fragment_dcc_ticketing_certificate_selection), AutoInject {
+
+    @Inject lateinit var viewModelFactory: CWAViewModelFactoryProvider.Factory
 
     private val args by navArgs<DccTicketingCertificateSelectionFragmentArgs>()
+    private val binding: FragmentDccTicketingCertificateSelectionBinding by viewBinding()
+    private val viewModel: DccTicketingCertificateSelectionViewModel by cwaViewModelsAssisted(
+        factoryProducer = { viewModelFactory },
+        constructorCall = { factory, _ ->
+            factory as DccTicketingCertificateSelectionViewModel.Factory
+            factory.create(args.transactionContext)
+        }
+    )
+    private val certificatesAdapter = DccTicketingCertificateSelectionAdapter()
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        binding.apply {
+            toolbar.setNavigationOnClickListener { returnToScreenWhereScannerWasOpened() }
+            recyclerView.apply {
+                adapter = certificatesAdapter
+                addItemDecoration(TopBottomPaddingDecorator(topPadding = R.dimen.spacing_tiny))
+            }
+            viewModel.events.observe(viewLifecycleOwner) { event ->
+                when (event) {
+                    is NavigateToConsentTwoFragment -> doNavigate(
+                        DccTicketingCertificateSelectionFragmentDirections
+                            .actionDccTicketingCertificateSelectionFragmentToDccTicketingConsentTwoFragment(
+                                event.transactionContext,
+                                event.selectedCertificateContainerId
+                            )
+                    )
+                }
+            }
+        }
+
+        lifecycleScope.launchWhenStarted {
+            val certificatesList = viewModel.getCertificates()
+            certificatesAdapter.update(certificatesList)
```

Comment:
Can we use live data instead and do the logic in the viewmodel? we pass the TransactionContext to the viewmodel as argument and you can in the  ViewModel  filter the certifcates and publish them to the fragment using some live data

```kotlin 
 class DccTicketingCertificateSelectionViewModel @AssistedInject constructor(
    .....
) : CWAViewModel(dispatcherProvider) {
    val certificats = liveData<Set<CwaCovidCertificate>> {
        emit(dccTicketingCertificateFilter.filter(transactionContext.accessTokenPayload?.vc))
    }
```